File: libraries/node/src/sources/osu/parse.ts
/**
 * Parser for .osu text.
 */
import type { OsuBeatmap, OsuHitObject, OsuTimingPoint } from "./model.js";
import { UrcError } from "../../error.js";

const METADATA_FIELDS: Record<string, "title" | "titleUnicode" | "artist" | "artistUnicode" | "creator" | "version"> = {
	Title: "title",
	TitleUnicode: "titleUnicode",
	Artist: "artist",
	ArtistUnicode: "artistUnicode",
	Creator: "creator",
	Version: "version"
};

/**
 * Parses .osu text into a source model.
 */
export function parseOsu(text: string): OsuBeatmap {
	const beatmap: OsuBeatmap = {
		mode: 3,
		title: null,
		titleUnicode: null,
		artist: null,
		artistUnicode: null,
		creator: null,
		version: null,
		circleSize: null,
		overallDifficulty: null,
		timingPoints: [],
		hitObjects: []
	};
	let section: string | null = null;
	const body = text.startsWith("﻿") ? text.slice(1) : text;

	for (const [offset, raw] of body.replace(/\r\n?/g, "\n").split("\n").entries()) {
		const lineNo = offset + 1;
		const line = raw.trim();

		if (line.startsWith("[") && line.endsWith("]")) {
			section = line.slice(1, -1);
			continue;
		}
		if (line === "" || line.startsWith("//") || section === null)
			continue;

		switch (section) {
			case "General":
				parseGeneral(beatmap, line, lineNo);
				break;
			case "Metadata":
				parseMetadata(beatmap, line);
				break;
			case "Difficulty":
				parseDifficulty(beatmap, line, lineNo);
				break;
			case "TimingPoints":
				beatmap.timingPoints.push(parseTimingPoint(line, lineNo));
				break;
			case "HitObjects":
				beatmap.hitObjects.push(parseHitObject(line, lineNo));
				break;
		}
	}

	return beatmap;
}

function parseGeneral(beatmap: OsuBeatmap, line: string, lineNo: number): void {
	const [key, value] = splitField(line);
	if (key !== undefined && value !== undefined && key === "Mode")
		beatmap.mode = toInt(value, lineNo, "Mode");

}

function parseMetadata(beatmap: OsuBeatmap, line: string): void {
	const [key, value] = splitField(line);
	const field = key === undefined || value === undefined ? undefined : METADATA_FIELDS[key];
	if (key !== undefined && value !== undefined && field !== undefined)
		beatmap[field] = value;

}

function parseDifficulty(beatmap: OsuBeatmap, line: string, lineNo: number): void {
	const [key, value] = splitField(line);
	if (key === undefined || value === undefined)
		return;

	if (key === "CircleSize")
		beatmap.circleSize = toFloat(value, lineNo, "CircleSize");
	else if (key === "OverallDifficulty")
		beatmap.overallDifficulty = toFloat(value, lineNo, "OverallDifficulty");

}

function parseTimingPoint(line: string, lineNo: number): OsuTimingPoint {
	const fields = line.split(",").map(field => field.trim());
	if (fields.length < 2)
		throw new UrcError("syntax", lineNo, `timing point needs at least 2 fields: ${line}`);

	return {
		time: toInt(fields[0], lineNo, "timing time"),
		beatLength: toFloat(fields[1], lineNo, "beat length"),
		meter: fields.length > 2 && fields[2] !== "" ? toInt(fields[2], lineNo, "meter") : 4,
		uninherited:
			fields.length > 6 && fields[6] !== "" ? toInt(fields[6], lineNo, "uninherited") !== 0 : true
	};
}

function parseHitObject(line: string, lineNo: number): OsuHitObject {
	const fields = line.split(",").map(field => field.trim());
	if (fields.length < 5)
		throw new UrcError("syntax", lineNo, `hit object needs at least 5 fields: ${line}`);

	const x = toInt(fields[0], lineNo, "hit object x");
	const time = toInt(fields[2], lineNo, "hit object time");
	const typeBits = toInt(fields[3], lineNo, "hit object type");

	const isHold = (typeBits & 128) !== 0;
	if (!isHold && (typeBits & 1) === 0)
		throw new UrcError("syntax", lineNo, `unsupported hit object type: ${typeBits}`);

	let endTime: number | null = null;
	if (isHold) {
		if (fields.length < 6)
			throw new UrcError("syntax", lineNo, `hold note needs an end time: ${line}`);

		endTime = toInt(fields[5].split(":", 1)[0].trim(), lineNo, "hold end time");
	}

	return { x, time, isHold, endTime };
}

function splitField(line: string): [string | undefined, string | undefined] {
	const separator = line.indexOf(":");
	if (separator === -1)
		return [undefined, undefined];

	return [line.slice(0, separator).trim(), line.slice(separator + 1).trim()];
}

function toInt(token: string, lineNo: number, label: string): number {
	const value = toFloat(token, lineNo, label);
	return Math.round(value);
}

function toFloat(token: string, lineNo: number, label: string): number {
	if (token.trim() === "")
		throw new UrcError("syntax", lineNo, `invalid ${label}: ${token}`);

	const value = Number(token);
	if (Number.isNaN(value))
		throw new UrcError("syntax", lineNo, `invalid ${label}: ${token}`);

	return value;
}

File: libraries/rust/src/sources/osu/parse.rs
//! Parser for `.osu` text.

use crate::error::{Result, UrcError};

use super::model::{OsuBeatmap, OsuHitObject, OsuTimingPoint};

/// Parses `.osu` text into a source model.
pub fn parse_osu(text: &str) -> Result<OsuBeatmap> {
    let mut beatmap = OsuBeatmap::default();
    let mut section: Option<String> = None;

    for (offset, raw) in text
        .strip_prefix('\u{feff}')
        .unwrap_or(text)
        .lines()
        .enumerate()
    {
        let line_no = offset as u32 + 1;
        let line = raw.trim();

        if line.starts_with('[') && line.ends_with(']') {
            section = Some(line[1..line.len() - 1].to_owned());
            continue;
        }

        if line.is_empty() || line.starts_with("//") || section.is_none() {
            continue;
        }

        match section.as_deref() {
            Some("General") => general(&mut beatmap, line, line_no)?,
            Some("Metadata") => metadata(&mut beatmap, line),
            Some("Difficulty") => difficulty(&mut beatmap, line, line_no)?,
            Some("TimingPoints") => beatmap.timing_points.push(timing_point(line, line_no)?),
            Some("HitObjects") => beatmap.hit_objects.push(hit_object(line, line_no)?),
            _ => {}
        }
    }

    Ok(beatmap)
}

fn general(beatmap: &mut OsuBeatmap, line: &str, line_no: u32) -> Result<()> {
    if let Some((key, value)) = line.split_once(':')
        && key.trim() == "Mode"
    {
        beatmap.mode = to_int(value.trim(), line_no, "Mode")?;
    }
    Ok(())
}

fn metadata(beatmap: &mut OsuBeatmap, line: &str) {
    if let Some((key, value)) = line.split_once(':') {
        let value = value.trim().to_owned();
        match key.trim() {
            "Title" => beatmap.title = Some(value),
            "TitleUnicode" => beatmap.title_unicode = Some(value),
            "Artist" => beatmap.artist = Some(value),
            "ArtistUnicode" => beatmap.artist_unicode = Some(value),
            "Creator" => beatmap.creator = Some(value),
            "Version" => beatmap.version = Some(value),
            _ => {}
        }
    }
}

fn difficulty(beatmap: &mut OsuBeatmap, line: &str, line_no: u32) -> Result<()> {
    if let Some((key, value)) = line.split_once(':') {
        match key.trim() {
            "CircleSize" => {
                beatmap.circle_size = Some(to_float(value.trim(), line_no, "CircleSize")?);
            }
            "OverallDifficulty" => {
                beatmap.overall_difficulty =
                    Some(to_float(value.trim(), line_no, "OverallDifficulty")?);
            }
            _ => {}
        }
    }
    Ok(())
}

fn timing_point(line: &str, line_no: u32) -> Result<OsuTimingPoint> {
    let fields: Vec<&str> = line.split(',').map(str::trim).collect();
    if fields.len() < 2 {
        return Err(UrcError::new(
            "syntax",
            line_no,
            format!("timing point needs at least 2 fields: {line:?}"),
        ));
    }

    let meter = match fields.get(2).filter(|field| !field.is_empty()) {
        Some(field) => to_int(field, line_no, "meter")? as u64,
        None => 4,
    };
    let uninherited = match fields.get(6).filter(|field| !field.is_empty()) {
        Some(field) => to_int(field, line_no, "uninherited")? != 0,
        None => true,
    };

    Ok(OsuTimingPoint {
        time: to_int(fields[0], line_no, "timing time")?,
        beat_length: to_float(fields[1], line_no, "beat length")?,
        meter,
        uninherited,
    })
}

fn hit_object(line: &str, line_no: u32) -> Result<OsuHitObject> {
    let fields: Vec<&str> = line.split(',').map(str::trim).collect();
    if fields.len() < 5 {
        return Err(UrcError::new(
            "syntax",
            line_no,
            format!("hit object needs at least 5 fields: {line:?}"),
        ));
    }

    let x = to_int(fields[0], line_no, "hit object x")?;
    let time = to_int(fields[2], line_no, "hit object time")?;
    let type_bits = to_int(fields[3], line_no, "hit object type")?;

    let is_hold = type_bits & 128 != 0;
    if !is_hold && type_bits & 1 == 0 {
        return Err(UrcError::new(
            "syntax",
            line_no,
            format!("unsupported hit object type: {type_bits}"),
        ));
    }

    let end_time = if is_hold {
        let token = fields.get(5).ok_or_else(|| {
            UrcError::new(
                "syntax",
                line_no,
                format!("hold note needs an end time: {line:?}"),
            )
        })?;
        Some(to_int(
            token.split(':').next().unwrap_or_default().trim(),
            line_no,
            "hold end time",
        )?)
    } else {
        None
    };

    Ok(OsuHitObject {
        x,
        time,
        is_hold,
        end_time,
    })
}

fn to_int(token: &str, line_no: u32, label: &str) -> Result<i64> {
    token
        .parse::<f64>()
        .map(|value| value.round() as i64)
        .map_err(|_| UrcError::new("syntax", line_no, format!("invalid {label}: {token:?}")))
}

fn to_float(token: &str, line_no: u32, label: &str) -> Result<f64> {
    token
        .parse::<f64>()
        .map_err(|_| UrcError::new("syntax", line_no, format!("invalid {label}: {token:?}")))
}
